File: server/app/linker.py
import re
import unicodedata
from collections import Counter
from pathlib import Path

from .config import settings
from .db import get_db

_WORD_RE = re.compile(r"\b[\wÁ-ú]+\b")
_plural_re = re.compile(r"^(?P<word>.+?)(?P<s>s|es)$")

_STOPWORDS = {
    "el", "la", "los", "las", "de", "del", "a", "al", "y", "o", "u", "en", "con",
    "que", "es", "un", "una", "para", "por", "se", "su", "sus", "lo", "le", "les",
    "como", "mas", "pero", "no", "ni", "esto", "este", "esta", "ese", "esa", "los",
}


def normalize(text: str) -> str:
    """Minúsculas, sin acentos/tildes, y con la 's' final de plural quitada."""
    text = text.lower()
    text = unicodedata.normalize("NFKD", text)
    text = "".join(c for c in text if not unicodedata.combining(c))
    words = []
    for w in _WORD_RE.findall(text):
        m = _plural_re.match(w)
        if m and len(m.group("word")) > 2:
            w = m.group("word")
        words.append(w)
    return " ".join(words)


def expand_variants(title: str, aliases: list[str]) -> list[str]:
    """Título + aliases + prefijos (primeras 1..N palabras) como variantes normalizadas."""
    out: list[str] = []
    seen: set[str] = set()

    def add(t: str):
        n = normalize(t)
        if n and n not in seen:
            seen.add(n)
            out.append(n)

    for s in [title, *aliases]:
        words = _WORD_RE.findall(s.lower())
        for i in range(1, len(words) + 1):
            add(" ".join(words[:i]))
    return out


def _load_notes(db) -> list[dict]:
    rows = db.execute(
        """
        SELECT d.id, d.title,
               COALESCE((SELECT GROUP_CONCAT(a.alias, '||') FROM aliases a
                         WHERE a.doc_id = d.id), '') AS aliases
        FROM docs d
        """
    ).fetchall()
    notes = []
    for r in rows:
        aliases = [a for a in (r["aliases"] or "").split("||") if a]
        notes.append({"id": r["id"], "title": r["title"], "aliases": aliases})
    return notes


def _tokenize_with_positions(text: str) -> list[dict]:
    """Devuelve tokens con start/end en char-offsets del texto original (case-preservado)."""
    tokens = []
    for m in _WORD_RE.finditer(text):
        tokens.append({"word": m.group(0), "start": m.start(), "end": m.end()})
    return tokens


def _build_span_index(text: str):
    """Mapa normalize(tokens) -> lista de spans (word literal, start, end) en el texto original."""
    index: dict[str, list[dict]] = {}
    for tok in _tokenize_with_positions(text):
        n = normalize(tok["word"])
        index.setdefault(n, []).append(
            {"lit": tok["word"], "start": tok["start"], "end": tok["end"]}
        )
    return index


def _match_variant_in_content(content: str, variant_norm: str):
    """Busca en content (string original) la secuencia de palabras cuya versión
    normalizada = variant_norm. Devuelve (literal_match, start, end) o None."""
    parts = variant_norm.split()
    tokens = _tokenize_with_positions(content)
    keys = [normalize(t["word"]) for t in tokens]
    for i in range(len(tokens) - len(parts) + 1):
        if keys[i : i + len(parts)] == parts:
            span = tokens[i : i + len(parts)]
            return (
                " ".join(t["word"] for t in span),
                span[0]["start"],
                span[-1]["end"],
            )
    return None


def run_linker(db=None, apply_auto: bool = True, target: str = "crdt") -> dict:
    db = db or get_db()
    notes = _load_notes(db)
    results = {"auto": [], "proposals": [], "notes": len(notes)}
    assert target in ("crdt", "mirror")

    read = _read_markdown if target == "crdt" else _read_mirror
    write = _apply_markdown if target == "crdt" else _write_mirror

    # Precompute per-note variants
    notes_data = []
    for n in notes:
        variants = expand_variants(n["title"], n["aliases"])
        exact_variants = {normalize(s) for s in [n["title"], *n["aliases"]] if normalize(s)}
        notes_data.append(
            {"id": n["id"], "title": n["title"], "variants": variants, "exact_variants": exact_variants}
        )

    # Variant -> notes that claim it (for ambiguity)
    variant_owner: dict[str, list[int]] = {}
    for idx, nd in enumerate(notes_data):
        for v in nd["variants"]:
            variant_owner.setdefault(v, []).append(idx)

    for note in notes_data:
        content = read(note["id"])
        if not content.strip():
            continue
        new_content = _link_doc(
            content, note, notes_data, variant_owner, db, apply_auto
        )
        if new_content is not None and new_content != content:
            if apply_auto:
                write(note["id"], new_content)
            results["auto"].append(
                {"doc_id": note["id"], "links": sorted(_extract_wikilinks(new_content) - _extract_wikilinks(content))}
            )

    results["proposals"] = _pending_proposals(db)
    return results


def _link_doc(content, note, notes_data, variant_owner, db, apply_auto):
    used_spans: list[tuple[int, int, str]] = []
    changed = False
    for other in notes_data:
        if other["id"] == note["id"]:
            continue
        for variant in other["variants"]:
            match = _match_variant_in_content(content, variant)
            if match is None:
                continue
            literal, start, end = match
            if any(start < e and s < end for s, e, _ in used_spans):
                continue  # overlaps an already-linked span
            owners = variant_owner.get(variant, [])
            exact = variant in other["exact_variants"]
            unambiguous = len(owners) == 1
            wikilink = f"[[{other['title']}]]"
            if wikilink in content:
                continue
            conf = 1.0 if exact and unambiguous else 0.6 if exact else 0.4
            if conf >= 0.999 and apply_auto:
                content = content[:start] + wikilink + content[end:]
                used_spans.append((start, start + len(wikilink), other["id"]))
                changed = True
                _record_proposal(db, note["id"], other["id"], literal, 1.0, status="applied")
            else:
                _record_proposal(db, note["id"], other["id"], literal, conf, status="pending")
    return content if changed else None


def _write_mirror(doc_id: str, content: str) -> None:
    target: Path = settings.mirror_dir / f"{doc_id}.md"
    if target.exists() and target.read_text(encoding="utf-8") == content:
        return
    tmp = target.with_suffix(".md.tmp")
    tmp.write_text(content, encoding="utf-8")
    tmp.replace(target)


def _read_mirror(doc_id: str) -> str:
    target: Path = settings.mirror_dir / f"{doc_id}.md"
    if target.exists():
        return target.read_text(encoding="utf-8")
    return ""


def _apply_markdown(doc_id: str, content: str) -> None:
    """Escribe el contenido al CRDT (fuente canónica) vía y-sweet.

    El mirror loop recogerá después este texto, así los links persisten aunque
    el CRDT se edite desde Obsidian.
    """
    from pycrdt import Doc, Text

    from .ysweet import ysweet_manager

    current = ysweet_manager().get_doc_as_update(doc_id)
    doc = Doc()
    if current:
        doc.apply_update(current)
    text = doc.get("content", type=Text)
    text.clear()
    text.insert(0, content)
    ysweet_manager().update_doc(doc_id, doc.get_update())


def _read_markdown(doc_id: str) -> str:
    """Lee el contenido markdown desde el CRDT (fuente canónica)."""
    from pycrdt import Doc, Text

    from .ysweet import ysweet_manager

    update = ysweet_manager().get_doc_as_update(doc_id)
    if not update:
        return ""
    doc = Doc()
    doc.apply_update(update)
    return str(doc.get("content", type=Text))


def _apply_link(doc_id: str, alias: str, wikilink: str) -> bool:
    """Sustituye una mención (alias) por un wikilink en el CRDT. Devuelve True si cambió algo."""
    content = _read_markdown(doc_id)
    if wikilink in content:
        return False
    token = _match_literal(content, alias)
    if token is None:
        return False
    new_content = content.replace(token, wikilink, 1)
    _apply_markdown(doc_id, new_content)
    return True


def _match_literal(content: str, alias: str) -> str | None:
    """Devuelve la subcadena literal de content que coincide con alias (normalizado,
    tolerante a acentos/plural). None si no hay match."""
    variant = normalize(alias)
    match = _match_variant_in_content(content, variant)
    if match is None:
        return None
    return match[0]


def _record_proposal(db, doc_id, target_id, alias, confidence, status="pending"):
    db.execute(
        """
        INSERT INTO linker_proposals (doc_id, target_id, alias, confidence, reason, status)
        VALUES (?, ?, ?, ?, 'auto-detected', ?)
        ON CONFLICT(doc_id, target_id, alias)
        DO UPDATE SET confidence = excluded.confidence, status = excluded.status
        """,
        (doc_id, target_id, alias, confidence, status),
    )
    db.commit()


def _pending_proposals(db) -> list[dict]:
    rows = db.execute(
        """
        SELECT p.rowid AS id, p.doc_id, p.target_id, p.alias, p.confidence, p.reason, p.status,
               d.title AS target_title
        FROM linker_proposals p JOIN docs d ON d.id = p.target_id
        WHERE p.status = 'pending'
        ORDER BY p.confidence DESC
        """
    ).fetchall()
    return [dict(r) for r in rows]


def _extract_wikilinks(content: str) -> set:
    return set(re.findall(r"\[\[([^\]]+)\]\]", content))


def extract_keywords(text: str, top_n: int = 10) -> list[str]:
    """keyBERT para extraer keywords; fallback TF-IDF/frecuencia simple."""
    text = (text or "").strip()
    if not text:
        return []
    try:
        from keybert import KeyBERT

        kw = KeyBERT(model="all-MiniLM-L6-v2")
        kws = kw.extract_keywords(text, keyphrase_ngram_range=(1, 2), stop_words="spanish", top_n=top_n)
        return [k for k, _ in kws]
    except Exception:
        return _freq_fallback(text, top_n)


def _freq_fallback(text: str, top_n: int) -> list[str]:
    words = [
        w
        for w in _WORD_RE.findall(text.lower())
        if w not in _STOPWORDS and len(w) > 3 and not w.isdigit()
    ]
    return [w for w, _ in Counter(words).most_common(top_n)]

File: crates/tg-server/src/linker.rs
//! Integración del linker puro (`tg-linker`) con el CRDT y la BD: equivalente a
//! `server/app/linker.py`.

use std::collections::BTreeSet;

use rusqlite::{params, Connection};
use serde_json::json;
use tg_linker::keywords::keywords_tfidf;
use tg_linker::{link_doc, LinkEdit, NoteSpec};

use crate::error::{ApiError, ApiResult};
use crate::state::AppState;

pub fn load_notes(db: &Connection) -> ApiResult<Vec<NoteSpec>> {
    let mut stmt = db.prepare(
        "SELECT d.id, d.title,
                COALESCE((SELECT GROUP_CONCAT(a.alias, '||') FROM aliases a
                          WHERE a.doc_id = d.id), '')
         FROM docs d",
    )?;
    let rows = stmt
        .query_map([], |r| {
            let aliases_raw: String = r.get(2)?;
            let aliases: Vec<String> = aliases_raw
                .split("||")
                .filter(|a| !a.is_empty())
                .map(String::from)
                .collect();
            Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?, aliases))
        })?
        .collect::<Result<Vec<_>, _>>()
        .map_err(ApiError::from)?;
    Ok(rows
        .into_iter()
        .map(|(id, title, aliases)| NoteSpec { id, title, aliases })
        .collect())
}

/// Ejecuta el linker sobre todas las notas. `target` = "crdt" (default) | "mirror".
pub async fn run_linker(
    st: &AppState,
    target: &str,
    user_id: &str,
) -> ApiResult<serde_json::Value> {
    if target != "crdt" && target != "mirror" {
        return Err(ApiError::bad_request("target must be 'crdt' or 'mirror'"));
    }
    let db = st.db.lock().await;
    let notes = load_notes(&db)?;

    let allowed: BTreeSet<String> = db
        .prepare("SELECT doc_id FROM doc_access WHERE user_id = ?1")?
        .query_map(params![user_id], |r| r.get::<_, String>(0))?
        .collect::<Result<BTreeSet<_>, _>>()
        .map_err(ApiError::from)?;
    let auto_filter = allowed.clone();

    let mut auto_out = Vec::new();
    for note in &notes {
        let content = match target {
            "crdt" => {
                let doc = st.crdt.open_or_create(&note.id).await.map_err(as_api)?;
                doc.read_text().await
            }
            _ => read_mirror(&st.settings.mirror_dir, &note.id),
        };
        if content.trim().is_empty() {
            continue;
        }
        let res = link_doc(&content, &note.id, &notes);
        for p in &res.proposals {
            record_proposal(
                &db,
                p.doc_id.as_str(),
                &p.target_id,
                &p.literal,
                p.confidence,
                "pending",
            )?;
        }
        for a in &res.auto {
            record_proposal(&db, note.id.as_str(), &a.target_id, "auto", 1.0, "applied")?;
        }

        if res.auto.is_empty() {
            continue;
        }
        match target {
            "crdt" => {
                let doc = st.crdt.open_or_create(&note.id).await.map_err(as_api)?;
                doc.apply_edits(res.auto.iter().map(|a| a.edit.clone()).collect())
                    .await
                    .map_err(as_api)?;
            }
            _ => {
                let mut new_content = content.clone();
                apply_edits_inplace(
                    &mut new_content,
                    &res.auto.iter().map(|a| a.edit.clone()).collect::<Vec<_>>(),
                );
                write_mirror(&st.settings.mirror_dir, &note.id, &new_content);
            }
        }
        let added: Vec<String> = res
            .auto
            .iter()
            .map(|a| a.edit.replacement.clone())
            .filter(|r| r.starts_with("[["))
            .collect::<BTreeSet<_>>()
            .into_iter()
            .collect();
        auto_out.push(json!({ "doc_id": note.id, "links": added }));
    }

    let pending = pending_proposals(&db)?
        .into_iter()
        .filter(|p| allowed.contains(p["doc_id"].as_str().unwrap_or("")))
        .collect::<Vec<_>>();

    Ok(json!({
        "auto": auto_out.into_iter().filter(|a| auto_filter.contains(a["doc_id"].as_str().unwrap_or(""))).collect::<Vec<_>>(),
        "proposals": pending,
        "notes": notes.len(),
    }))
}

/// Aplica edits en orden descendente sobre un String (offsets byte del original).
pub fn apply_edits_inplace(content: &mut String, edits: &[LinkEdit]) {
    let mut sorted = edits.to_vec();
    sorted.sort_by_key(|e| std::cmp::Reverse(e.start));
    for e in &sorted {
        content.replace_range(e.start..e.end, &e.replacement);
    }
}

pub fn read_mirror(mirror_dir: &std::path::Path, doc_id: &str) -> String {
    let path = mirror_dir.join(format!("{doc_id}.md"));
    std::fs::read_to_string(path).unwrap_or_default()
}

pub fn write_mirror(mirror_dir: &std::path::Path, doc_id: &str, content: &str) {
    let target = mirror_dir.join(format!("{doc_id}.md"));
    if target.exists() && std::fs::read_to_string(&target).is_ok_and(|c| c == content) {
        return;
    }
    let _ = std::fs::create_dir_all(mirror_dir);
    let tmp = target.with_extension("tmp");
    let _ = std::fs::write(&tmp, content);
    let _ = std::fs::rename(&tmp, &target);
}

pub fn record_proposal(
    db: &Connection,
    doc_id: &str,
    target_id: &str,
    alias: &str,
    confidence: f64,
    status: &str,
) -> ApiResult<()> {
    db.execute(
        "INSERT INTO linker_proposals (doc_id, target_id, alias, confidence, reason, status)
         VALUES (?1, ?2, ?3, ?4, 'auto-detected', ?5)
         ON CONFLICT(doc_id, target_id, alias)
         DO UPDATE SET confidence = excluded.confidence, status = excluded.status",
        params![doc_id, target_id, alias, confidence, status],
    )?;
    Ok(())
}

pub fn pending_proposals(db: &Connection) -> ApiResult<Vec<serde_json::Value>> {
    let mut stmt = db.prepare(
        "SELECT p.rowid AS id, p.doc_id, p.target_id, p.alias, p.confidence, p.reason, p.status,
                d.title AS target_title
         FROM linker_proposals p JOIN docs d ON d.id = p.target_id
         WHERE p.status = 'pending'
         ORDER BY p.confidence DESC",
    )?;
    let rows = stmt
        .query_map([], |r| {
            Ok(json!({
                "id": r.get::<_, i64>("id")?,
                "doc_id": r.get::<_, String>("doc_id")?,
                "target_id": r.get::<_, String>("target_id")?,
                "alias": r.get::<_, String>("alias")?,
                "confidence": r.get::<_, f64>("confidence")?,
                "reason": r.get::<_, String>("reason")?,
                "status": r.get::<_, String>("status")?,
                "target_title": r.get::<_, String>("target_title")?,
            }))
        })?
        .collect::<Result<Vec<_>, _>>()
        .map_err(ApiError::from)?;
    Ok(rows)
}

/// Contenido a efectos de keywords/índice: preferencia al CRDT si tiene texto, si no
/// al espejo (docs poblados por mirror/import antes de conectarse al CRDT).
pub async fn read_best(st: &AppState, doc_id: &str) -> String {
    if let Some(doc) = st.crdt.get(doc_id).await {
        let t = doc.read_text().await;
        if !t.trim().is_empty() {
            return t;
        }
    }
    read_mirror(&st.settings.mirror_dir, doc_id)
}

/// Keywords deterministas TF-IDF (corpus = textos a los que el usuario tiene acceso).
pub async fn keywords_for_doc(
    st: &AppState,
    doc_id: &str,
    user_id: &str,
    top_n: usize,
) -> ApiResult<Vec<String>> {
    let db = st.db.lock().await;
    let owns: bool = db
        .query_row(
            "SELECT 1 FROM doc_access WHERE doc_id=?1 AND user_id=?2",
            params![doc_id, user_id],
            |_| Ok(()),
        )
        .is_ok();
    if !owns {
        return Err(ApiError::forbidden("no access to this doc"));
    }
    let corpus_ids: Vec<String> = db
        .prepare("SELECT doc_id FROM doc_access WHERE user_id=?1")?
        .query_map(params![user_id], |r| r.get(0))?
        .collect::<Result<_, _>>()
        .map_err(ApiError::from)?;
    drop(db);

    let text = read_best(st, doc_id).await;
    if text.trim().is_empty() {
        return Ok(Vec::new());
    }
    let mut corpus: Vec<String> = Vec::new();
    for id in corpus_ids {
        let t = read_best(st, &id).await;
        if !t.trim().is_empty() {
            corpus.push(t);
        }
    }
    let refs: Vec<&str> = corpus.iter().map(|s| s.as_str()).collect();
    Ok(keywords_tfidf(&text, &refs, top_n))
}

fn as_api(e: anyhow::Error) -> ApiError {
    ApiError::from(e)
}

/// Parsea los wikilinks `[[...]]` de un contenido, en orden de aparición.
pub fn extract_wikilinks(content: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut rest = content;
    while let Some(idx) = rest.find("[[") {
        let after = &rest[idx + 2..];
        if let Some(close) = after.find("]]") {
            out.push(after[..close].to_string());
            rest = &after[close + 2..];
        } else {
            break;
        }
    }
    out
}

/// Localiza la primera mención que coincide con `alias` (normalizado, tolerante a
/// acentos/plural) y devuelve su rango en offsets de byte. Paridad con `_match_literal`.
pub fn match_literal_offset(content: &str, alias: &str) -> Option<(usize, usize)> {
    use tg_linker::normalize::{normalize, normalize_word, tokenize_with_positions};
    let variant = normalize(alias);
    let parts: Vec<&str> = variant.split(' ').filter(|s| !s.is_empty()).collect();
    if parts.is_empty() {
        return None;
    }
    let tokens = tokenize_with_positions(content);
    let keys: Vec<String> = tokens.iter().map(|t| normalize_word(t.lit)).collect();
    if parts.len() > keys.len() {
        return None;
    }
    for win in 0..=keys.len() - parts.len() {
        let mut ok = true;
        for (k, p) in parts.iter().enumerate() {
            if keys[win + k] != *p {
                ok = false;
                break;
            }
        }
        if ok {
            return Some((tokens[win].start, tokens[win + parts.len() - 1].end));
        }
    }
    None
}
